servers/services/src/main/java/org/projectnessie/services/impl/TreeApiImpl.java

```diff
@@ -893,14 +893,23 @@ public <R> R getEntries(
         }
       }
 
-      Predicate<KeyEntry> filterPredicate = filterEntries(filter);
+      // apply filter as early as possible to avoid work (i.e. content loading, authz checks)
+      // for entries that we will eventually throw away
+      final int namespaceFilterDepth = namespaceDepth == null ? 0 : namespaceDepth.intValue();
+      if (namespaceFilterDepth > 0) {
+        Predicate<ContentKey> depthFilter = e -> e.getElementCount() >= namespaceFilterDepth;
+        contentKeyPredicate =
+            contentKeyPredicate == null ? depthFilter : contentKeyPredicate.and(depthFilter);
+      }
+      final Predicate<KeyEntry> filterPredicate = filterEntries(filter);
+      final Predicate<KeyEntry> loadContentPredicate = withContent ? filterPredicate : null;
 
       try (PaginationIterator<KeyEntry> entries =
           getStore()
               .getKeys(
                   refWithHash.getHash(),
                   pagingToken,
-                  withContent,
+                  loadContentPredicate,
```

Comment:
WDYT about adding `Predicate<KeyEntry>` to `KeyRestrictions` as `keyEntryPredicate`? Then, "with content" can remain `boolean` and the authZ iterator does not have to change.

Granted, `VersionStore` will have two layers of filtering (one based on `ContentKey`, the other on `KeyEntry`) but it seems to yield simpler code at the service impl. layer :thinking:

Reply:
I had a similar idea that I shared with @XN137 . That said I don't think the new predicate should be declared inside `KeyRestrictions` because it only applies to `getKeys` (not to `getDiffs`).

Also `KeyEntry` is probably not ideal for that because it needs to materialize the content in order to be instantiated. We need another type that basically exposes only `ContentKey` and `Content.Type`.

Reply:
I agree, `KeyRestrictions` is not the right place for the new filter.

Reply:
`KeyEntry` could be instantiated without content, but I'd be fine with introducing a more specific type for the sake of clarity.